stl/inc/algorithm

```diff
@@ -394,6 +793,194 @@ _NODISCARD pair<_FwdIt1, _FwdIt2> mismatch(
+            }
+        }
+        // clang-format on
+    };
+
+    inline constexpr _Mismatch_fn mismatch{_Not_quite_object::_Construct_tag{}};
+
+    // VARIABLE ranges::equal
+    class _Equal_fn : private _Not_quite_object {
+    private:
+        template <class _It1, class _It2, class _Size, class _Pr, class _Pj1, class _Pj2>
+        _NODISCARD static constexpr bool _Equal_count(
+            _It1 _First1, _It2 _First2, _Size _Count, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            for (; _Count != 0; ++_First1, (void) ++_First2, --_Count) {
+                if (!_STD _C_invoke(_Pred, _STD _C_invoke(_Proj1, *_First1), _STD _C_invoke(_Proj2, *_First2))) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        template <class _It1, class _Se1, class _It2, class _Se2, class _Pr, class _Pj1, class _Pj2>
+        _NODISCARD static constexpr bool _Equal_4(
+            _It1 _First1, _Se1 _Last1, _It2 _First2, _Se2 _Last2, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            for (;;) {
+                if (_First1 == _Last1) {
+                    return _First2 == _Last2;
+                } else if (_First2 == _Last2) {
+                    return false;
+                }
+
+                if (!_STD _C_invoke(_Pred, _STD _C_invoke(_Proj1, *_First1), _STD _C_invoke(_Proj2, *_First2))) {
+                    return false;
+                }
+
+                ++_First1;
+                ++_First2;
+            }
+        }
+
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <input_iterator _It1, sentinel_for<_It1> _Se1, input_iterator _It2, sentinel_for<_It2> _Se2,
+            class _Pr = ranges::equal_to, class _Pj1 = identity, class _Pj2 = identity>
+            requires indirectly_comparable<_It1, _It2, _Pr, _Pj1, _Pj2>
+        _NODISCARD constexpr bool operator()(_It1 _First1, _Se1 _Last1, _It2 _First2, _Se2 _Last2, _Pr _Pred = {},
+            _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            _Adl_verify_range(_First1, _Last1);
+            _Adl_verify_range(_First2, _Last2);
+            auto _UFirst1      = _Get_unwrapped(_STD move(_First1));
+            const auto _ULast1 = _Get_unwrapped(_STD move(_Last1));
+            auto _UFirst2      = _Get_unwrapped(_STD move(_First2));
+            const auto _ULast2 = _Get_unwrapped(_STD move(_Last2));
+
+            if constexpr (sized_sentinel_for<_Se1, _It1> && sized_sentinel_for<_Se2, _It2>) {
+                const auto _Count = _ULast1 - _UFirst1;
+                if (_Count != _ULast2 - _UFirst2) {
+                    return false;
+                }
+
+                return _Equal_count(_STD move(_UFirst1), _STD move(_UFirst2), _Count,
+                    _Pass_fn(_Pred), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            } else {
+                return _Equal_4(_STD move(_UFirst1), _ULast1, _STD move(_UFirst2), _ULast2,

```

Comment:
Moving iterators is important because single-pass iterators might be move-only. Sentinels, on the other hand, must always be copyable. So this is _correct_ but potentially less performant and likely more confusing than moving the sentinels would be due to the asymmetry.

So I suppose my final answer here is "Yes, we should conventionally move sentinels when possible, despite that we need not necessarily do so."